### PR TITLE
[WorkflowCard] Implement WorkflowCard component

### PR DESCRIPTION
**Pull request by manojleaton**
Fixes #BLUI-5097 .

#### Changes proposed in this Pull Request:

- Added WorkflowCard component
- Updated JS docs
- Added test cases
- Updated the demo in example
- Updated the Readme

#### Screenshots / Screen Recording (if applicable)
Screenshot 2024-01-26 at 9 35 32 PM
Screenshot 2024-01-26 at 9 51 31 PM

#### To Test:

- yarn start:example-ios/android

#### Any specific feedback you are looking for?

- Need feedback on rendering svg in the card background, as it is done differently in ReactNative. We have added svg to the background in React workflow.
- For loading do we need to use ProgressBar component?

### CODE REVIEW

**Comment by huayunh** [comment]
what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

File: login-workflow/src/__tests__/components/WorkflowCard/WorkflowCard-test.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * @format
+ */
+
```
Comment (huayunh) [inline login-workflow/src/__tests__/components/WorkflowCard/WorkflowCard-test.tsx:4]: I didn't check the rest of the code... does this comment do anything?

**Review by huayunh — COMMENTED** [review]
what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

File: login-workflow/src/components/WorkflowCard/WorkflowCard.tsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    loading?: boolean;
+    /**
+     * A custom background to render behind the card
+     */
+    backgroundImage?: string;
+};
+
+type WorkflowCardHeaderProps = CardTitleProps;
+
+type WorkflowCardActionsProps = CardActionsProps;
```
Comment (surajeaton) [inline login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:30]: Please export this type.

File: login-workflow/src/components/WorkflowCard/WorkflowCard.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+type WorkflowCardBaseProps = ViewProps & {
+    /**
+     * If true, a blocking progress spinner will be displayed over the card
+     */
+    loading?: boolean;
+    /**
+     * A custom background to render behind the card
+     */
+    backgroundImage?: string;
```
Comment (surajeaton) [inline login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:25]: We are not using this `backgroundImage` prop anywhere. We should use that.

File: login-workflow/src/components/WorkflowCard/WorkflowCard.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ * @param backgroundImage to display card background
+ *
+ * @category Component
+ */
+
+type WorkflowCardBaseProps = ViewProps & {
+    /**
+     * If true, a blocking progress spinner will be displayed over the card
+     */
+    loading?: boolean;
```
Comment (surajeaton) [inline login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:21]: We should make use of this prop in workflow card.

File: login-workflow/src/__tests__/components/WorkflowCard/WorkflowCard-test.tsx (hunk 1 above)
Comment (manojleaton) [reply login-workflow/src/__tests__/components/WorkflowCard/WorkflowCard-test.tsx:4]: Removed the comment.

**Comment by manojleaton** [comment]
> what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

1. Created a issue for the header being chopped off in Tablet in React Native Component Library,  [Issue](https://github.com/etn-ccis/blui-react-native-component-library/issues/531).
2. Removed the default border radius of the RNP Card component. Please check the screenshot for the updated mobile view.
Screenshot 2024-01-29 at 6 01 42 PM

**Comment by huayunh** [comment]
Sorry I just noticed this -- what's happening with the text field? The red underline is on the top?

**Comment by joebochill** [comment]
This looks broken when the device is rotated:
image

Hardcoding fixed dimension for the card should probably come with a maxWidth/maxHeight too just to be on the safe side.

Looks like you are always rendering a SafeAreaView. Per conversation I had with Arshdeep, we should be using a normal view and conditionally adding top padding depending on the presence of a WorkflowCardHeader child. Are you going to add that here, or is @ArshdeepSingh3-eaton going to make the changes in a follow-up?

**Comment by joebochill** [comment]
> Sorry I just noticed this -- what's happening with the text field? The red underline is on the top?

Looks like maybe an issue with the WorkflowCardBody? If I remove it, the input looks fine.

**Comment by manojleaton** [comment]
> This looks broken when the device is rotated: image
> 
> Hardcoding fixed dimension for the card should probably come with a maxWidth/maxHeight too just to be on the safe side.
> 
> Looks like you are always rendering a SafeAreaView. Per conversation I had with Arshdeep, we should be using a normal view and conditionally adding top padding depending on the presence of a WorkflowCardHeader child. Are you going to add that here, or is @ArshdeepSingh3-eaton going to make the changes in a follow-up?

1. I have added maxHeight and maxWidth to the Card, now it's not breaking. 
Screenshot 2024-01-31 at 4 34 54 PM


2. I have changed to normal View component and added top padding code in this PR itself as @ArshdeepSingh3-eaton suggested.

**Comment by huayunh** [comment]
> This looks broken when the device is rotated:

This raises an interesting issue -- do we really want users to use auth workflow in landscape on a phone? Portrait on a phone and portrait/landscape on a tablet makes sense to me. But landscape mobile login is something I'd not support. I'd lock the device orientation on mobile to portrait.

**Comment by huayunh** [comment]
And this -- the workflow card is currently avoiding the safe area on the bottom:
image

I would have expected that white area to touch the device edge, and that iPhone home slider thing to turn black.

**Comment by ArshdeepSingh3-eaton** [comment]
can you add style prop as well

**Comment by ArshdeepSingh3-eaton** [comment]
@huayunh i don't think that locking the landscape mode on mobile will help as when this library is used by an end user and they have not locked landscape mode
it could cause issue

File: login-workflow/src/hooks/useScreenDimensions.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/**
+ * @packageDocumentation
+ * @module Hooks
+ * @internal
+ */
+import { useWindowDimensions } from 'react-native';
+
+type ScreenWidthProps = {
```
Comment (surajeaton) [inline login-workflow/src/hooks/useScreenDimensions.ts:8]: Please rename type to ScreenDimensionsProps

File: login-workflow/src/hooks/useScreenDimensions.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    width: number;
```
Comment (surajeaton) [inline login-workflow/src/hooks/useScreenDimensions.ts:9]: Also add js doc

File: login-workflow/src/components/WorkflowCard/WorkflowCard.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        },
+        tablet: {
+            height: MAX_CARD_HEIGHT,
+            width: MAX_CARD_WIDTH,
+            borderRadius: theme.roundness * 2,
+            overflow: 'hidden',
+        },
+    });
+
+export type WorkflowCardHeaderProps = CardTitleProps;
```
Comment (surajeaton) [inline login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:58]: Please move all the type definitions to types.ts file

**Review by ektaghag-eaton — APPROVED** [review]
Working fine.

**Pull request merged**